clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,224 @@
+                                    snippet(cx, receiver.span, "..")
+                                ),
+                                Applicability::Unspecified,
+                            );
+                            diag.span_suggestion(first_local.span, "remove this line", "", Applicability::Unspecified);
+                            if !r.extra_locals.is_empty() {
+                                let extra_local_suggestions = r
+                                    .extra_locals
+                                    .iter()
+                                    .map(|x| {
+                                        (
+                                            x.init.unwrap().span,
+                                            snippet(cx, first_local.pat.span, "..").to_string(),
+                                        )
+                                    })
+                                    .collect::<Vec<_>>();
+
+                                diag.multipart_suggestion(
+                                    "initialize this variable to be the `Some` variant (may need dereferencing)",
+                                    extra_local_suggestions,
+                                    Applicability::Unspecified,
+                                );
+                            }
+                            if !r.extra_exprs.is_empty() {
+                                let index_accesses = r
+                                    .extra_exprs
+                                    .iter()
+                                    .map(|x| (x.span, snippet(cx, first_local.pat.span, "..").to_string()))
+                                    .collect::<Vec<_>>();
+
+                                diag.multipart_suggestion(
+                                    "set this index to be the `Some` variant (may need dereferencing)",
+                                    index_accesses,
+                                    Applicability::Unspecified,
+                                );
+                            }
+                        } else {
+                            let mut index_accesses = vec![(
+                                if_expr.cond.span,
+                                format!("let Some(element) = {}.first()", snippet(cx, receiver.span, "..")),
+                            )];
+                            index_accesses.extend(r.extra_exprs.iter().map(|x| (x.span, "element".to_owned())));
+
+                            diag.multipart_suggestion(
+                                "consider using if..let syntax (variable may need to be dereferenced)",
+                                index_accesses,
+                                Applicability::Unspecified,
+                            );
+                        }
+                    },
+                );
+            }
+        }
+    }
+}
+
+struct IndexCheckResult<'a> {
+    // the receiver for the index operation
+    pub index_receiver: Option<&'a Expr<'a>>,
```

Comment:
`index_receiver` has some additional bounds on it, for example it is only `Some` if the index is a primitive constant, and not some kind of dynamic value stored in a variable or constant somewhere else, so I think it can stay. I'll update the comment on the field in the struct to better explain this.

Reply:
Although, on closer inspection, this is addressed by one of your earlier requested changes, so this can probably actually be removed.